File: deploy/production/import-tools.js
#!/usr/bin/env node
/**
 * 批量导入工具到生产环境
 * 用法：node deploy/production/import-tools.js
 * 无外部依赖，只用 Node.js 内置 http 模块
 */
const http = require('http');
const fs = require('fs');
const path = require('path');

const HOST = '127.0.0.1';
const PORT = 3000;

// 读取工具数据
const dataPath = path.join(__dirname, 'tools-data.json');
if (!fs.existsSync(dataPath)) {
  console.error('❌ tools-data.json 不存在，请确保文件在 deploy/production/ 目录下');
  process.exit(1);
}
const tools = JSON.parse(fs.readFileSync(dataPath, 'utf-8'));
console.log(`📦 共 ${tools.length} 件工具待导入\n`);

// HTTP 请求封装
function request(method, path, body) {
  return new Promise((resolve, reject) => {
    const data = body ? JSON.stringify(body) : null;
    const opts = {
      hostname: HOST, port: PORT, method,
      path,
      headers: { 'Content-Type': 'application/json' }
    };
    if (data) opts.headers['Content-Length'] = Buffer.byteLength(data);
    const req = http.request(opts, (res) => {
      let chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        const text = Buffer.concat(chunks).toString();
        try {
          resolve({ status: res.statusCode, data: JSON.parse(text) });
        } catch {
          resolve({ status: res.statusCode, data: text });
        }
      });
    });
    req.on('error', reject);
    req.setTimeout(10000, () => { req.destroy(); reject(new Error('timeout')); });
    if (data) req.write(data);
    req.end();
  });
}

async function login() {
  console.log('🔐 登录...');
  const res = await request('POST', '/api/auth/login', { username: 'admin', password: '123456' });
  if (res.status !== 200 || !res.data.access_token) {
    throw new Error('登录失败: ' + JSON.stringify(res.data));
  }
  return res.data.access_token;
}

async function apiCall(token, method, path, body) {
  const headers = { 'Authorization': `Bearer ${token}` };
  return new Promise((resolve, reject) => {
    const data = body ? JSON.stringify(body) : null;
    const opts = {
      hostname: HOST, port: PORT, method,
      path,
      headers: { ...headers, 'Content-Type': 'application/json' }
    };
    if (data) opts.headers['Content-Length'] = Buffer.byteLength(data);
    const req = http.request(opts, (res) => {
      let chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => {
        const text = Buffer.concat(chunks).toString();
        try {
          resolve({ status: res.statusCode, data: JSON.parse(text) });
        } catch {
          resolve({ status: res.statusCode, data: text });
        }
      });
    });
    req.on('error', reject);
    req.setTimeout(10000, () => { req.destroy(); reject(new Error('timeout')); });
    if (data) req.write(data);
    req.end();
  });
}

async function ensureEntity(token, getPath, postPath, nameField, codeField, name, code, extras = {}) {
  // 先查列表
  const listRes = await apiCall(token, 'GET', getPath);
  if (listRes.status === 200) {
    const items = Array.isArray(listRes.data) ? listRes.data : (listRes.data.tools || listRes.data.items || []);
    const found = items.find(item => item[nameField] === name);
    if (found) {
      return found.id || found.entity_id || found[Object.keys(found).find(k => k.endsWith('_id'))];
    }
  }
  // 不存在则创建
  const body = { [nameField]: name, [codeField]: code, ...extras };
  const res = await apiCall(token, 'POST', postPath, body);
  if (res.status === 200 || res.status === 201) {
    const id = res.data.id || res.data.tool_id || res.data[Object.keys(res.data).find(k => k.endsWith('_id'))];
    console.log(`  ✅ 创建 ${nameField}: ${name} (id=${id})`);
    return id;
  }
  console.log(`  ⚠️  ${name} -> ${res.status}: ${JSON.stringify(res.data).substring(0, 100)}`);
  return null;
}

async function main() {
  try {
    const token = await login();
    console.log('  ✅ 登录成功\n');

    // Category mapping from the data
    const catNames = [...new Set(tools.map(t => t.category_name))];
    const catMap = {};
    for (const cn of catNames) {
      const listRes = await apiCall(token, 'GET', '/api/tool-categories');
      const cats = Array.isArray(listRes.data) ? listRes.data : [];
      const found = cats.find(c => c.category_name === cn);
      if (found) {
        catMap[cn] = found.category_id;
        console.log(`  📂 ${cn} (id=${found.category_id})`);
      }
    }
    console.log();

    // Warehouse
    const whList = await apiCall(token, 'GET', '/api/warehouses');
    const whs = Array.isArray(whList.data) ? whList.data : [];
    let mainWh = whs.find(w => w.warehouse_name === '主仓库');
    if (!mainWh) {
      const cr = await apiCall(token, 'POST', '/api/warehouses', { warehouse_name: '主仓库', warehouse_code: '主仓库' });
      mainWh = cr.data;
      console.log(`  🏭 创建主仓库 (id=${mainWh.warehouse_id})`);
    } else {
      console.log(`  🏭 主仓库 (id=${mainWh.warehouse_id})`);
    }
    const mainWhId = mainWh.warehouse_id;

    // Shelves
    const shelfNames = [...new Set(tools.map(t => t.shelf_name))];
    const shelfMap = {};
    for (const sn of shelfNames) {
      const sc = sn.replace('号货架', ''); // "1号货架" -> "HJ-01"
      const id = await ensureEntity(token, '/api/shelves', '/api/shelves', 'shelf_name', 'shelf_code', sn, `HJ-${sc.padStart(2, '0')}`, { warehouse_id: mainWhId });
      if (id) shelfMap[sn] = id;
    }
    console.log();

    // Storage locations
    const locNames = [...new Set(tools.map(t => t.location_name).filter(Boolean))];
    const locMap = {};
    for (const ln of locNames) {
      const shelfName = ln.endsWith('层') ? ln.substring(0, ln.length - 2) + '货架' : ln;
      if (shelfName && shelfName.includes('号货架')) {
        const shelfNum = shelfName.charAt(0);
        const layerNum = ln.charAt(ln.length - 2);
        const lc = `HJ${shelfNum.padStart(2, '0')}-${layerNum.padStart(2, '0')}`;
        const sid = shelfMap[shelfName];
        const id = await ensureEntity(token, '/api/storage-locations', '/api/storage-locations', 'location_name', 'location_code', ln, lc, { warehouse_id: mainWhId, shelf_id: sid });
        if (id) locMap[ln] = id;
      }
    }
    console.log();

    // Import tools
    console.log('='.repeat(55));
    console.log(`🔧 批量导入 ${tools.length} 件工具...`);
    console.log('='.repeat(55));

    let ok = 0, skip = 0, fail = 0;
    const errs = [];

    for (let i = 0; i < tools.length; i++) {
      const t = tools[i];
      const body = {
        tool_code: t.tool_code,
        tool_name: t.tool_name,
        category_id: catMap[t.category_name],
        warehouse_id: mainWhId,
        shelf_id: shelfMap[t.shelf_name] || null,
        storage_location_id: locMap[t.location_name] || null,
        status: 'available',
        description: t.description || '',
      };
      const res = await apiCall(token, 'POST', '/api/tools', body);
      if (res.status === 200 || res.status === 201) {
        ok++;
      } else if (res.status === 400 && res.data && res.data.message && res.data.message.includes('已存在')) {
        skip++;
      } else {
        fail++;
        errs.push(`${t.tool_code} ${t.tool_name}`);
      }

      if ((i + 1) % 25 === 0 || i === tools.length - 1) {
        console.log(`  [${String(i+1).padStart(3)}/${tools.length}] ✅${ok} ⏭️${skip} ❌${fail}`);
      }
    }

    console.log(`\n${'='.repeat(55)}`);
    console.log(`📊 导入完成: ${ok} 新建, ${skip} 已存在, ${fail} 失败`);
    if (errs.length > 0) {
      console.log(`\n失败明细 (前10):`);
      errs.slice(0, 10).forEach(e => console.log(`  ❌ ${e}`));
    }
  } catch (err) {
    console.error('💥 错误:', err.message);
    process.exit(1);
  }
}

main();

File: import_tools.py
#!/usr/bin/env python
"""批量导入工具到工器具管理系统 — 通过 API"""
import requests
import sys
import openpyxl

BASE = "http://localhost:3000/api"
TOKEN = None

def api(method, path, data=None):
    headers = {"Authorization": f"Bearer {TOKEN}"} if TOKEN else {}
    url = f"{BASE}{path}"
    try:
        if method == "GET":
            r = requests.get(url, headers=headers, timeout=10)
        elif method == "POST":
            r = requests.post(url, json=data, headers=headers, timeout=10)
        else:
            raise ValueError(f"Unknown method: {method}")
    except requests.exceptions.ConnectionError:
        print(f"  ❌ 连接失败: {url} (后端未启动?)")
        sys.exit(1)
    if r.status_code >= 400:
        print(f"  ❌ {method} {path} -> {r.status_code}: {r.text[:200]}")
        return None
    return r.json()

# ===== 1. Login =====
print("🔐 登录...")
resp = api("POST", "/auth/login", {"username": "admin", "password": "123456"})
if not resp or "access_token" not in resp:
    print("❌ 登录失败")
    sys.exit(1)
TOKEN = resp["access_token"]
print(f"  ✅ admin 登录成功\n")

# ===== 2. Load Excel =====
print("📖 读取 Excel...")
wb = openpyxl.load_workbook("工具录入清单_待补充.xlsx")
ws = wb.active
tools = []
for row in ws.iter_rows(min_row=2, values_only=True):
    if row[0] and str(row[0]).startswith("G-"):
        tools.append({
            "tool_code": str(row[0]).strip(),
            "tool_name": str(row[1]).strip() if row[1] else "",
            "cat":       str(row[2]).strip() if row[2] else "",
            "wh":        str(row[3]).strip() if row[3] else "主仓库",
            "shelf":     str(row[4]).strip() if row[4] else "",
            "loc":       str(row[5]).strip() if row[5] else "",
        })
print(f"  ✅ {len(tools)} 条记录\n")

# ===== 3. Categories =====
print("📂 分类...")
CAT_CODES = {"电动工具": "DDGJ", "手动工具": "SDGJ", "仪表": "YB"}
existing = api("GET", "/tool-categories") or []
cat_map = {c["category_name"]: c["category_id"] for c in existing}
for cn, cc in CAT_CODES.items():
    if cn in cat_map:
        print(f"  ✓ {cn} (id={cat_map[cn]})")
        continue
    r = api("POST", "/tool-categories", {"category_name": cn, "category_code": cc})
    if r:
        cat_map[cn] = r["category_id"]
        print(f"  ✅ 新建: {cn} [{cc}] (id={cat_map[cn]})")

# ===== 4. Warehouse =====
print("\n🏭 仓库...")
existing = api("GET", "/warehouses") or []
wh_map = {w["warehouse_name"]: w["warehouse_id"] for w in existing}
needed = set(t["wh"] for t in tools)
for wn in needed:
    if wn in wh_map:
        print(f"  ✓ {wn} (id={wh_map[wn]})")
        continue
    r = api("POST", "/warehouses", {"warehouse_name": wn, "warehouse_code": wn})
    if r:
        wh_map[wn] = r["warehouse_id"]
        print(f"  ✅ 新建: {wn} (id={wh_map[wn]})")

MAIN_WH_ID = wh_map.get("主仓库")

# ===== 5. Shelves =====
print("\n📦 货架...")
SHELF_MAP = {"1号货架": "HJ-01", "2号货架": "HJ-02", "3号货架": "HJ-03"}
existing = api("GET", "/shelves") or []
shelf_map = {s["shelf_name"]: s["shelf_id"] for s in existing}
needed = set(t["shelf"] for t in tools)
for sn in needed:
    if sn in shelf_map:
        print(f"  ✓ {sn} (id={shelf_map[sn]})")
        continue
    sc = SHELF_MAP.get(sn, sn)
    r = api("POST", "/shelves", {"shelf_name": sn, "shelf_code": sc, "warehouse_id": MAIN_WH_ID})
    if r:
        shelf_map[sn] = r["shelf_id"]
        print(f"  ✅ 新建: {sn} [{sc}] (id={shelf_map[sn]})")

# ===== 6. Storage Locations =====
print("\n📍 货位...")
existing = api("GET", "/storage-locations") or []
loc_map = {l["location_name"]: l["location_id"] for l in existing}
needed = sorted(set(t["loc"] for t in tools))

def loc_code(ln):
    """'1号货架-1层' -> 'HJ01-01'"""
    parts = ln.split("-")
    shelf_num = parts[0][0]  # "1"
    layer_num = parts[-1][0] if parts[-1][0].isdigit() else parts[-1]
    return f"HJ{shelf_num.zfill(2)}-{str(layer_num).zfill(2)}"

for ln in needed:
    if ln in loc_map:
        print(f"  ✓ {ln} (id={loc_map[ln]})")
        continue
    shelf_name = ln.rsplit("-", 1)[0]  # "1号货架-1层" -> "1号货架"
    sid = shelf_map.get(shelf_name)
    lc = loc_code(ln)
    r = api("POST", "/storage-locations", {
        "location_name": ln, "location_code": lc,
        "warehouse_id": MAIN_WH_ID, "shelf_id": sid
    })
    if r:
        loc_map[ln] = r["location_id"]
        print(f"  ✅ 新建: {ln} [{lc}] (id={loc_map[ln]})")
    else:
        print(f"  ❌ 失败: {ln}")

# ===== 7. Import Tools =====
print(f"\n{'='*55}")
print(f"🔧 批量导入 {len(tools)} 件工具...")
print(f"{'='*55}")

ok = 0
fail = 0
errs = []

for i, t in enumerate(tools):
    data = {
        "tool_code": t["tool_code"],
        "tool_name": t["tool_name"],
        "category_id": cat_map.get(t["cat"]),
        "warehouse_id": MAIN_WH_ID,
        "shelf_id": shelf_map.get(t["shelf"]),
        "storage_location_id": loc_map.get(t["loc"]),
        "status": "available",
    }
    r = api("POST", "/tools", data)
    if r and "tool_id" in r:
        ok += 1
    else:
        fail += 1
        errs.append(f"{t['tool_code']} {t['tool_name']}")

    if (i + 1) % 25 == 0:
        print(f"  [{i+1:3d}/{len(tools)}] ✅{ok} ❌{fail}")

# final line
print(f"  [{len(tools)}/{len(tools)}] ✅{ok} ❌{fail}")

print(f"\n{'='*55}")
print(f"📊 导入完成: {ok}/{len(tools)} 成功" + (f", {fail} 失败" if fail else ""))
if errs:
    print(f"\n失败明细 (前10):")
    for e in errs[:10]:
        print(f"  ❌ {e}")
    if len(errs) > 10:
        print(f"  ... 及其他 {len(errs)-10} 条")
